src/odemis/driver/actuator.py

```diff
@@ -760,10 +760,119 @@ def stop(self, axes=None):
+
+class Convert3DStage(ConvertStage):
+    """
+    Extends original ConvertStage with an additional axis Z
+    """
+    def __init__(self, name, role, dependencies, axes,
+                 rotation=(0, 0, 0), scale=None, translation=None, **kwargs):
+        """
+        dependencies (dict str -> actuator): name to objective lens actuator
+        axes (list of 3 strings): names of the axes for x, y and z
+        scale (None tuple of 3 floats): scale factor from exported to original position
+        rotation (tuple of 3 floats): rotation factor rz, ry, rx (in radians)
+        translation (None or tuple of 3 floats): translation offset (in m)
+        """
+        assert len(axes) == 3
+        if len(dependencies) != 1:
+            raise ValueError("ConvertStage needs 1 dependency")
+
+        self._dependency = list(dependencies.values())[0]
+        self._axes_dep = {"x": axes[0], "y": axes[1], "z": axes[2]}
+        if scale is None:
+            scale = (1, 1, 1)
+        if translation is None:
+            translation = (0, 0, 0)
+        if rotation.count(0) < 2:
+            raise ValueError("Convert3DStage allows only one rotation angle to be > 0.")
+
+        axes_def = {"x": self._dependency.axes[axes[0]],
+                    "y": self._dependency.axes[axes[1]],
+                    "z": self._dependency.axes[axes[2]]}
+        model.Actuator.__init__(self, name, role, dependencies=dependencies, axes=axes_def, **kwargs)
 
+        self._metadata[model.MD_POS_COR] = translation
+        self._metadata[model.MD_ROTATION_COR] = rotation
+        self._metadata[model.MD_PIXEL_SIZE_COR] = scale
+        self._updateConversion()
+
+        # RO, as to modify it the client must use .moveRel() or .moveAbs()
+        self.position = model.VigilantAttribute({"x": 0, "y": 0, "z": 0},
+                                                unit="m", readonly=True)
+        # it's just a conversion from the dep's position
+        self._dependency.position.subscribe(self._updatePosition, init=True)
+
+        if model.hasVA(self._dependency, "referenced"):
+            self.referenced = model.VigilantAttribute({}, readonly=True)
+            self._dependency.referenced.subscribe(self._updateReferenced, init=True)
+
+        if model.hasVA(self._dependency, "speed"):
+            speed_axes = set(self._dependency.speed.value.keys())
+            if set(axes) <= speed_axes:
+                self.speed = model.VigilantAttribute({}, readonly=True)
+                self._dependency.speed.subscribe(self._updateSpeed, init=True)
+            else:
+                logging.info("Axes %s of dependency are missing from .speed, so not providing it",
+                             set(axes) - speed_axes)
+
+    def _get_rot_matrix(self, invert=False):
+        # Overrides parent class method with rotation matrices for the 3 angles
+        # NB: Rotation is counterclockwise
+        # TODO: handle rotation for the 3 angles at the same time
```

Comment:
Shouldn't the dot matrix multiplication be enough? What currently doesn't work with multiple angles?

Reply:
I tried to use more than 1 angle at the same time, the issue I have is that the convert stage doesn't reach the intended position. I have tried to use the rotation matrix as it is, as well as trying to invert the order this way, but had the same issue:
```
 if invert:
    rx_mat @ ry_mat @ rz_mat
```
	
Btw, the issue I was referring to in our talk that could be related is called Gimbal lock and it's described here:
https://omar-shehata.medium.com/how-to-fix-gimbal-lock-in-n-dimensions-f2f7baec2b5e

Reply:
Strange. I've just tried, and for me it looked like it works.
```
      if invert:
            rotation = tuple(r*-1 for r in rotation)
:
      if invert: 
          return rx_mat @ ry_mat @ rz_mat
      else:
          return rz_mat @ ry_mat @ rx_mat
```

Even if this doesn't work, I'd suggest to leave it like that, because that should be the closest to the "correct way".